go/private/actions/archive.bzl

```diff
@@ -88,13 +90,13 @@ def emit_archive(ctx, go_toolchain, mode=None, golib=None, goembed=None, direct=
   for src in source.asm:
     obj = ctx.actions.declare_file("{}/{}.o".format(out_dir, src.basename[:-2]))
     go_toolchain.actions.asm(ctx, go_toolchain, mode=mode, source=src, hdrs=source.headers, out_obj=obj)
-    extra_objects += [obj]
+    extra_objects.append(obj)
   archive = goembed.cgo_info.archive if goembed.cgo_info else None
 
   transitive = direct
   for a in direct:
     if a.mode != mode: fail("Archive mode does not match {} is {} expected {}".format(a.library.label, mode_string(a.mode), mode_string(mode)))
-    transitive += a.transitive
+    transitive = sets.union(transitive, a.transitive)
```

Comment:
I think if we're setting Bazel 0.7.0 as the minimum, the new depset constructor is supported (with direct and transitive parameters). That would be more efficient here. My earlier attempt was in the first commit of #1026.

The depset constructor is O(m+n). Building a depset incrementally like this will be O(m*n).

Reply:
I thought that the cost of (a + b + c) was the same as ((a + b) + c) if all of a, b and c were depsets?
either way, I would prefer to stick with sets.union and rely on it to be implemented correctly (currently it uses += to add depsets together underneath)

Reply:
What you said is true, but it's faster to build a unified depset in a single step from a list of depsets. You can accumulate a list of depsets in the loop, or use a list comprehension.

```
transitive = depset(direct = list(direct), transitive = [d.transitive for d in direct])
```

The issue is that when you use `.union` or `+`, Bazel tries to de-duplicate all the direct and nested elements, which means creating two `CompactHashSets`, and calling `hashCode` and `equals` for each. If one of the sets has a lot of direct elements, this happens for all of them on every union.

Reply:
I thought that Bazel only did de-duplication when a depset is collapsed back to a list rather than on construction?

Reply:
And even then, we ought to change union to use the new depset constructor rather than repeated +=
sets.union should *always* be the best answer for how to union a bunch of sets :)

Reply:
> I thought that Bazel only did de-duplication when a depset is collapsed back to a list rather than on construction?

Sadly, that is not the case. See the constructor in [SkylarkNestedSet.java](https://github.com/bazelbuild/bazel/blob/17cb02dc4dad4a6bad59bc08d9789a3a0e5a17fe/src/main/java/com/google/devtools/build/lib/syntax/SkylarkNestedSet.java#L119). I was looking through this yesterday to understand why it was hashing things added to the set. Nothing in here worked the way I expected.

Note that this only de-duplicates the combined list of direct values and the combined list of transitive depsets. It doesn't look at the contents of the transitive depsets.

> And even then, we ought to change union to use the new depset constructor rather than repeated +=
> sets.union should always be the best answer for how to union a bunch of sets :)

Agreed, but I'm not sure what their minimum supported Bazel version is. The depset constructor was new in Bazel 0.7 (the first revision of my PR yesterday broke on 0.6). In any case, sets.union will still be slower than necessary when it's called in a loop. It will be better to build a list of depsets, then call sets.union on all of them. It should be implemented like this:

```
def _union(*args):
  for i, arg in enumerate(args):
    if type(arg) != "depset":
      args[i] = depset(args)
  return depset(transitive = args)
```